src/Nethermind/Nethermind.Synchronization.Test/Trie/HealingTreeTests.cs

```diff
@@ -106,4 +119,102 @@ private void recovery_works<T>(
             action.Should().Throw<MissingTrieNodeException>();
         }
     }
+
+    [TestCase(INodeStorage.KeyScheme.Hash)]
+    [TestCase(INodeStorage.KeyScheme.HalfPath)]
+    public async Task HealingTreeTest(INodeStorage.KeyScheme keyScheme)
+    {
+        IContainer CreateNode()
+        {
+            ConfigProvider configProvider = new ConfigProvider();
+            configProvider.GetConfig<IPruningConfig>().Mode = PruningMode.Full;
+            configProvider.GetConfig<IInitConfig>().StateDbKeyScheme = keyScheme;
+            return new ContainerBuilder()
+                .AddModule(new TestNethermindModule(configProvider))
+                .AddSingleton<IBlockTree>(Build.A.BlockTree().OfChainLength(1).TestObject)
+                .OnBuild((ctx) =>
+                {
+                    ILogManager logManager = ctx.Resolve<ILogManager>();
+                    ISyncPeerPool peerPool = ctx.Resolve<ISyncPeerPool>();
+                    // Always recover flag
+                    ctx.Resolve<IWorldStateManager>().InitializeNetwork(
+                        new GetNodeDataTrieNodeRecovery(peerPool, logManager, true),
+                        new SnapTrieNodeRecovery(peerPool, logManager, true));
+                })
+                .Build();
+        }
+
+        await using IContainer server = CreateNode();
+        await using IContainer client = CreateNode();
+
+        IDb clientStateDb = client.ResolveNamed<IDb>(DbNames.State);
+        IDb serverStateDb = server.ResolveNamed<IDb>(DbNames.State);
+
+        // Add some data to the server.
+        Hash256 stateRoot;
+        {
+            IWorldState mainWorldState = server.Resolve<MainBlockProcessingContext>().WorldState;
+            mainWorldState.StateRoot = Keccak.EmptyTreeHash;
+
+            for (int i = 0; i < 100; i++)
+            {
+                Address address = new Address(Keccak.Compute(i.ToString()));
+                mainWorldState.CreateAccount(address, (UInt256)i, (UInt256)i);
+            }
+
+            Address storageAddress = new Address(Keccak.Compute("storage"));
+            mainWorldState.CreateAccount(storageAddress, 100, 100);
+            for (int i = 1; i < 100; i++)
+            {
+                mainWorldState.Set(new StorageCell(storageAddress, (UInt256)i), i.ToBigEndianByteArray());
+            }
+
+            mainWorldState.Commit(Cancun.Instance);
+            mainWorldState.CommitTree(1);
+
+            stateRoot = mainWorldState.StateRoot;
+        }
```

Comment:
This test is bit hard to follow, maybe move some parts to private methods with good naming?

Reply:
Done